src/hooks/useSearchQuery.jsx

```diff
@@ -0,0 +1,29 @@
+import { useState, useEffect, useCallback } from 'react';
+import { useNavigate, useSearchParams } from 'react-router-dom';
+
+export const useSearchQuery = () => {
+  const [query, setQuery] = useState('');
+  const navigate = useNavigate();
+  const [searchParams] = useSearchParams();
+
+  useEffect(() => {
+    const queryParam = searchParams.get('query') || '';
+    setQuery(queryParam);
+  }, [searchParams]);
+
+  const updateQuery = useCallback(
+    (newQuery, route = '') => {
+      const trimmedQuery = newQuery.trim();
+      if (trimmedQuery) {
+        searchParams.set('query', trimmedQuery);
+        navigate(`${route}?${searchParams.toString()}`);
```

Comment:
just in case, navigate can accept object, so you don't need to construct the whole URL by yourself

`navigate({ pathname: '/SOME_PATH_IN_APP', search: '?somekey=somevalue' })`

if you for some reason want to create search params by yourself, you can use https://reactrouter.com/6.28.1/utils/create-search-params